core/cog-gamepad-manette.c

```diff
@@ -0,0 +1,319 @@
+/*
+ * gamepad.h
+ * Copyright (C) 2022 Igalia S.L.
+ *
+ * Distributed under terms of the MIT license.
+ */
+
+#include "cog-gamepad.h"
+
+#include <libmanette.h>
+#include <linux/input-event-codes.h>
+#include <wpe/wpe.h>
+
+typedef struct _Provider Provider;
+typedef struct _Gamepad  Gamepad;
+
+struct _Provider {
+    struct wpe_gamepad_provider *wk_provider;
+    ManetteMonitor              *monitor;
+};
+
+struct _Gamepad {
+    struct wpe_gamepad *wk_gamepad;
+    ManetteDevice      *device;
+};
+
+static void *
+provider_create(struct wpe_gamepad_provider *wk_provider)
+{
+    Provider *provider = g_new0(Provider, 1);
+    provider->wk_provider = wk_provider;
+
+    return provider;
+}
+
+static void
+provider_destroy(void *data)
+{
+    Provider *provider = data;
+
+    if (!provider)
+        return;
+
+    g_clear_object(&provider->monitor);
+    g_free(provider);
+}
+
+static void
+add_device(Provider *provider, ManetteDevice *device)
+{
+    g_debug("gamepad: %s - connected", manette_device_get_name(device));
+
+    wpe_gamepad_provider_dispatch_gamepad_connected(provider->wk_provider, GPOINTER_TO_UINT(device));
+}
+
+static gboolean
+add_connected_devices(gpointer data)
+{
+    Provider                     *provider = data;
+    g_autoptr(ManetteMonitorIter) iter = manette_monitor_iterate(provider->monitor);
+    ManetteDevice                *device;
+
+    while (manette_monitor_iter_next(iter, &device)) {
+        add_device(provider, device);
+    };
+
+    return G_SOURCE_REMOVE;
+}
+
+static void
+device_connected_cb(ManetteMonitor *monitor, ManetteDevice *device, gpointer data)
+{
+    add_device(data, device);
+}
+
+static void
+device_disconnected_cb(ManetteMonitor *monitor, ManetteDevice *device, gpointer data)
+{
+    Provider *provider = data;
+
+    g_debug("gamepad: %s - disconnected", manette_device_get_name(device));
+
+    wpe_gamepad_provider_dispatch_gamepad_disconnected(provider->wk_provider, GPOINTER_TO_UINT(device));
+}
+
+static void
+provider_start(void *data)
+{
+    Provider *provider = data;
+
+    /* double call to start? */
+    if (provider->monitor)
+        return;
+
+    g_debug("gamepad: starting monitor");
+
+    provider->monitor = manette_monitor_new();
+    g_signal_connect(provider->monitor, "device-connected", G_CALLBACK(device_connected_cb), provider);
+    g_signal_connect(provider->monitor, "device-disconnected", G_CALLBACK(device_disconnected_cb), provider);
+
+    g_idle_add(add_connected_devices, provider);
+}
+
+static void
+provider_stop(void *data)
+{
+    Provider *provider = data;
+    int       n;
+
+    if (!provider)
+        return;
+
+    g_debug("gamepad: stoping monitor");
+
+    n = g_signal_handlers_disconnect_by_data(provider->monitor, provider);
+    g_assert(n == 2);
```

Comment:
I would remove this assertion and trust GLib to be able to do the right thing 😉 

Also this is going to cause warnings when building with assertions disabled, due to `n` being assigned but never read back again.